File: app/src/main/kotlin/com/foodshare/core/utilities/InputSanitizer.kt
package com.foodshare.core.utilities

import android.net.Uri

/**
 * Input sanitization utilities for security.
 *
 * Architecture (Frameo pattern):
 * - Local Kotlin implementations for text sanitization
 * - No JNI required for these pure string operations
 */
object InputSanitizer {

    // MARK: - Text Sanitization

    /**
     * Sanitize text input by normalizing whitespace and removing control characters.
     */
    fun sanitizeText(input: String): String {
        return input
            .trim()
            // Normalize newlines (CRLF and CR to LF)
            .replace("\r\n", "\n")
            .replace("\r", "\n")
            // Collapse multiple spaces to single space
            .replace(Regex("[ \\t]+"), " ")
            // Collapse multiple newlines to max 2
            .replace(Regex("\\n{3,}"), "\n\n")
            // Remove zero-width characters
            .replace(Regex("[\\u200B-\\u200D\\uFEFF]"), "")
    }

    /**
     * Escape HTML special characters to prevent XSS attacks.
     */
    fun escapeHTML(input: String): String {
        return input
            .replace("&", "&amp;")   // Must be first
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#39;")
    }

    /**
     * Strip all HTML tags from input.
     */
    fun stripHTML(input: String): String {
        return input
            // Remove script/style tags and their content
            .replace(Regex("<script[^>]*>[\\s\\S]*?</script>", RegexOption.IGNORE_CASE), "")
            .replace(Regex("<style[^>]*>[\\s\\S]*?</style>", RegexOption.IGNORE_CASE), "")
            // Remove all HTML tags
            .replace(Regex("<[^>]+>"), "")
            // Decode common HTML entities
            .replace("&nbsp;", " ")
            .replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&quot;", "\"")
            .replace("&#39;", "'")
            // Clean up whitespace
            .trim()
    }

    /**
     * Sanitize for SQL (basic protection - always use parameterized queries)
     */
    fun sanitizeForSQL(input: String): String {
        return input
            .replace("'", "''")
            .replace("\\", "\\\\")
    }

    // MARK: - URL Sanitization

    /**
     * Validate and sanitize URL string
     * @return Sanitized URL or null if invalid
     */
    fun sanitizeURL(input: String): String? {
        val trimmed = input.trim()

        return try {
            val uri = Uri.parse(trimmed)
            val scheme = uri.scheme?.lowercase()

            if (scheme == "http" || scheme == "https") {
                uri.toString()
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    // MARK: - Email Sanitization

    /**
     * Sanitize email address
     */
    fun sanitizeEmail(input: String): String {
        return input.trim().lowercase()
    }

    // MARK: - Filename Sanitization

    /**
     * Sanitize filename for safe storage
     */
    fun sanitizeFilename(input: String): String {
        // Remove path separators and dangerous characters
        val dangerousChars = setOf('/', '\\', ':', '*', '?', '"', '<', '>', '|')
        var sanitized = input.filter { it !in dangerousChars }

        // Remove leading dots (hidden files)
        sanitized = sanitized.trimStart('.')

        // Limit length
        if (sanitized.length > 255) {
            sanitized = sanitized.take(255)
        }

        // Ensure not empty
        if (sanitized.isEmpty()) {
            sanitized = "unnamed"
        }

        return sanitized
    }

    // MARK: - Length Validation

    /**
     * Truncate string to maximum length
     */
    fun truncate(input: String, maxLength: Int): String {
        return if (input.length <= maxLength) {
            input
        } else {
            input.take(maxLength)
        }
    }

    /**
     * Validate string length is within bounds
     */
    fun validateLength(input: String, min: Int = 0, max: Int = Int.MAX_VALUE): Boolean {
        return input.length in min..max
    }
}

/**
 * Validation patterns for common input types
 */
object ValidationPattern {
    /** Email regex pattern */
    val email = Regex("^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$")

    /** Phone number pattern (basic) */
    val phone = Regex("^[+]?[0-9]{10,15}$")

    /** Username pattern (alphanumeric + underscore) */
    val username = Regex("^[a-zA-Z0-9_]{3,30}$")

    /** Strong password pattern */
    val strongPassword = Regex("^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@\$!%*?&])[A-Za-z\\d@\$!%*?&]{8,}$")

    /**
     * Check if string matches pattern
     */
    fun matches(input: String, pattern: Regex): Boolean {
        return pattern.matches(input)
    }
}

// MARK: - String Extensions

/**
 * Sanitized version of the string
 */
val String.sanitized: String
    get() = InputSanitizer.sanitizeText(this)

/**
 * HTML-escaped version of the string
 */
val String.htmlEscaped: String
    get() = InputSanitizer.escapeHTML(this)

/**
 * HTML-stripped version of the string
 */
val String.htmlStripped: String
    get() = InputSanitizer.stripHTML(this)

/**
 * Truncated to max length
 */
fun String.truncated(maxLength: Int): String =
    InputSanitizer.truncate(this, maxLength)

/**
 * Check if string is within length bounds
 */
fun String.isWithinLength(min: Int = 0, max: Int = Int.MAX_VALUE): Boolean =
    InputSanitizer.validateLength(this, min, max)

File: Sources/FoodShare/Core/Utilities/InputSanitizer.swift
//
//  InputSanitizer.swift
//  Foodshare
//
//  Input sanitization utilities for security
//

import Foundation

// MARK: - Input Sanitizer

enum InputSanitizer {
    // MARK: - Text Sanitization

    /// Sanitize text input by removing dangerous characters
    static func sanitizeText(_ input: String) -> String {
        var sanitized = input

        // Remove null bytes
        sanitized = sanitized.replacingOccurrences(of: "\0", with: "")

        // Remove control characters (except newlines and tabs)
        sanitized = sanitized.unicodeScalars
            .filter { scalar in
                scalar == "\n" || scalar == "\t" || scalar == "\r" ||
                    !CharacterSet.controlCharacters.contains(scalar)
            }
            .map { String($0) }
            .joined()

        return sanitized
    }

    /// Sanitize HTML by escaping special characters
    static func escapeHTML(_ input: String) -> String {
        input
            .replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
            .replacingOccurrences(of: "\"", with: "&quot;")
            .replacingOccurrences(of: "'", with: "&#39;")
    }

    /// Strip all HTML tags from input
    static func stripHTML(_ input: String) -> String {
        input.replacingOccurrences(
            of: "<[^>]+>",
            with: "",
            options: .regularExpression,
        )
    }

    /// Sanitize for SQL (basic protection - use parameterized queries)
    static func sanitizeForSQL(_ input: String) -> String {
        input
            .replacingOccurrences(of: "'", with: "''")
            .replacingOccurrences(of: "\\", with: "\\\\")
    }

    // MARK: - URL Sanitization

    /// Validate and sanitize URL string
    static func sanitizeURL(_ input: String) -> String? {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)

        // Check for valid URL
        guard let url = URL(string: trimmed),
              let scheme = url.scheme?.lowercased(),
              ["http", "https"].contains(scheme) else {
            return nil
        }

        return url.absoluteString
    }

    // MARK: - Email Sanitization

    /// Sanitize email address
    static func sanitizeEmail(_ input: String) -> String {
        input
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
    }

    // MARK: - Filename Sanitization

    /// Sanitize filename for safe storage
    static func sanitizeFilename(_ input: String) -> String {
        // Remove path separators and dangerous characters
        let dangerous = CharacterSet(charactersIn: "/\\:*?\"<>|")
        let components = input.unicodeScalars.filter { !dangerous.contains($0) }
        var sanitized = String(String.UnicodeScalarView(components))

        // Remove leading dots (hidden files)
        while sanitized.hasPrefix(".") {
            sanitized.removeFirst()
        }

        // Limit length
        if sanitized.count > 255 {
            sanitized = String(sanitized.prefix(255))
        }

        // Ensure not empty
        if sanitized.isEmpty {
            sanitized = "unnamed"
        }

        return sanitized
    }

    // MARK: - Length Validation

    /// Truncate string to maximum length
    static func truncate(_ input: String, maxLength: Int) -> String {
        if input.count <= maxLength {
            return input
        }
        return String(input.prefix(maxLength))
    }

    /// Validate string length is within bounds
    static func validateLength(
        _ input: String,
        min: Int = 0,
        max: Int = Int.max,
    ) -> Bool {
        input.count >= min && input.count <= max
    }
}

// MARK: - String Extension

extension String {
    /// Sanitized version of the string
    var sanitized: String {
        InputSanitizer.sanitizeText(self)
    }

    /// HTML-escaped version of the string
    var htmlEscaped: String {
        InputSanitizer.escapeHTML(self)
    }

    /// HTML-stripped version of the string
    var htmlStripped: String {
        InputSanitizer.stripHTML(self)
    }

    /// Truncated to max length
    func truncated(to maxLength: Int) -> String {
        InputSanitizer.truncate(self, maxLength: maxLength)
    }

    /// Check if string is within length bounds
    func isWithinLength(min: Int = 0, max: Int = Int.max) -> Bool {
        InputSanitizer.validateLength(self, min: min, max: max)
    }
}

// MARK: - Validation Patterns

enum ValidationPattern {
    /// Email regex pattern
    static let email = "^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$"

    /// Phone number pattern (basic)
    static let phone = "^[+]?[0-9]{10,15}$"

    /// Username pattern (alphanumeric + underscore)
    static let username = "^[a-zA-Z0-9_]{3,30}$"

    /// Strong password pattern
    static let strongPassword = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&])[A-Za-z\\d@$!%*?&]{8,}$"

    /// Check if string matches pattern
    static func matches(_ input: String, pattern: String) -> Bool {
        input.range(of: pattern, options: .regularExpression) != nil
    }
}
